### PR TITLE
Throw exception on READER_IDLE and WRITER_IDLE states

### PR DESCRIPTION
**Pull request by AlessandroPatti**
Throws exceptions if one of these events gets fired, so that we can tell on which direction the connection was idling.

### CODE REVIEW

File: src/main/java/com/google/devtools/build/lib/remote/http/IdleTimeoutHandler.java [hunk 1]
```diff
@@ -29,20 +32,59 @@
...
+          closed = true;
+          break;
+        case READER_IDLE:
+          if (timeoutException == WriteTimeoutException.INSTANCE) {
+            ctx.fireExceptionCaught(WriteTimeoutReaderIdleException.INSTANCE);
+          }
+          else if (timeoutException == ReadTimeoutException.INSTANCE) {
+            ctx.fireExceptionCaught(ReadTimeoutReaderIdleException.INSTANCE);
+          }
+          break;
```
Comment (thelvis4) [inline src/main/java/com/google/devtools/build/lib/remote/http/IdleTimeoutHandler.java:76]: Don't we need to `ctx.close(); closed = true;` here too?

File: src/main/java/com/google/devtools/build/lib/remote/http/IdleTimeoutHandler.java [hunk 2]
```diff
@@ -29,20 +32,59 @@
...
+        case WRITER_IDLE:
+          if (timeoutException == WriteTimeoutException.INSTANCE) {
+            ctx.fireExceptionCaught(WriteTimeoutWriterIdleException.INSTANCE);
+          }
+          else if (timeoutException == ReadTimeoutException.INSTANCE) {
+            ctx.fireExceptionCaught(ReadTimeoutWriterIdleException.INSTANCE);
+          }
+          break;
```
Comment (thelvis4) [inline src/main/java/com/google/devtools/build/lib/remote/http/IdleTimeoutHandler.java:84]: ditto

File: src/main/java/com/google/devtools/build/lib/remote/http/IdleTimeoutHandler.java [hunk 3]
```diff
@@ -29,20 +32,59 @@
...
+
+  public final static class ReadTimeoutWriterIdleException extends IOException {
+    public static final ReadTimeoutWriterIdleException INSTANCE = new ReadTimeoutWriterIdleException();
+    private ReadTimeoutWriterIdleException() {}
+  }
+
   @SuppressWarnings("GoodTime-ApiWithNumericTimeUnit")
   public IdleTimeoutHandler(long timeoutSeconds, TimeoutException timeoutException) {
-    super(/* readerIdleTime= */ 0, /* writerIdleTime= */ 0, timeoutSeconds, TimeUnit.SECONDS);
+    super(timeoutSeconds, timeoutSeconds, timeoutSeconds, TimeUnit.SECONDS);
```
Comment (thelvis4) [inline src/main/java/com/google/devtools/build/lib/remote/http/IdleTimeoutHandler.java:57]: what's going to change if passing `timeoutSeconds` to `readerIdleTime` and `writerIdleTime`?

Comment (AlessandroPatti) [reply src/main/java/com/google/devtools/build/lib/remote/http/IdleTimeoutHandler.java:57]: It makes the corresponding events get fired when the timeout kicks in.

File: src/main/java/com/google/devtools/build/lib/remote/http/IdleTimeoutHandler.java (hunk 1 above)
Comment (AlessandroPatti) [reply src/main/java/com/google/devtools/build/lib/remote/http/IdleTimeoutHandler.java:76]: I wasn't sure on that and wanted to investigate it a bit first. Updated it.

**Review by thelvis4 — APPROVED** [review]

File: src/main/java/com/google/devtools/build/lib/remote/http/IdleTimeoutHandler.java (hunk 1 above)
Comment (rzsombor) [reply src/main/java/com/google/devtools/build/lib/remote/http/IdleTimeoutHandler.java:76]: I suggest a slightly different approach, we could just track the number of times reads and writes time out. Then write them out as part of the log message that tell us how many bytes were transmitted for which cas url.

To maintain behaviour the channel should only be closed when both read and write times out - that is, when `IdleState.ALL_IDLE` is happening. When `READER_IDLE` or `WRITER_IDLE` happens, we just increment 2 respective counters in this class. Then when `ALL_IDLE` hits us, we create a new TimeoutException instance that takes these 2 counters as properties.

At `HttpDownloadHandler` and `HttpUploadHandler`'s `exceptionCaught` callback these counters are propagated forward all the way down where logging is happening.

wdyt?

Comment (rzsombor) [reply src/main/java/com/google/devtools/build/lib/remote/http/IdleTimeoutHandler.java:76]: What I don't know is what the event flow is if we use the same timeout value for all 3 timeout params. So fo we reach ALL_IDLE by:
- READER_IDLE -> WRITER_IDLE -> ALL_IDLE or
- READER_IDLE -> ALL_IDLE.

Maybe we could use the submitted timeout -1 sec for reader and writer timeouts to make sure it's the former? :)

**Pull request merged**